### ISSUE TITLE
[question] best way to have my own reward model which is backed by rules

### ISSUE THREAD

**Issue by yananchen1989** [huggingface/trl#2518]
hi there,
i see that in recent versions from 0.11, especially in PPO, there is a big change that only `reward_model` of `nn.Module` is accepted for scoring. in contrast to [previous versions](https://huggingface.co/docs/trl/v0.11.4/en/ppo_trainer) that I can use my own reward module to get the scores on the fly. 

thus, my question is, if i have my own reward module, which is backed by rules instead of a LM, what is the recommended practice to integrate it into the current version ? 

what comes into my mind are two manners:

-  perhaps to change the function in `get_reward` https://github.com/huggingface/trl/blob/main/trl/trainer/utils.py#L1051 ?
-  Use my own reward module to score samples and train a module as the reward LM and then pass it into the ppo_trainer ?

My question also applied to `OnlineDPOTrainer`.

Thanks.

**Comment by oliveiraeliel** [comment]
Hi, I have the same question as you do.

I think that there must be some easy way to simply write a reward function as a `nn.Module`, so we don't have to refactor anything, but I didn't tried it yet.

But I also think that `PPOTrainer` should accept a `custom_get_reward_function` as an optional parameter. In this case, anyone could define its own reward function, and would be a clean solution.

**Comment by nityadav** [comment]
@yananchen1989 Thanks for posting this as I was stuck with a similar issue (but for `OnlineDPOTrainer`). The easiest workaround for me was to subclass the trainer class (`OnlineDPOTrainer`) and override the `training_step` with my custom `get_reward` logic, and rest of the implementation being the same as in the original method.

**Comment by yananchen1989** [comment]
correct me if i am wrong. 
would like to know the primary motivation to rewrite the dpo from older version to current trainer unified version. maybe for better efficiency ?

i understand that recent TRL versions wants to unify the pipeline in a more neat and organized manner across these different RL methods, where Trainer is the pivotal module. 
so for some methods like ppo the reward module is needed, it is also directed passed into the trainer. while for say dpo, sft, there is no provision for reward module. 

however. this could cause excessive encapsulation since it if hard to modularize the the reward module. 
the core reason is that in practical cases, reward module can be of any form not just a single torch.nn module which just score the whole output. the reward module may be a mixture and may be of dependence on external parameters, the prompt and most importantly it can not score the ppo trainer' outputs in a batch mode. 
anyway the flexibility is significantly reduced. 
although as you know the current unified pipeline is very fine with  other mothods such as dpo as they do not have the reward concerns and the reward module is implicitly expressed within the algorithm. 


in my view, there is no need to rigidly transfer these rl methods into a unified training framework.


pls advise.

**Comment by August-murr** [comment]
Ultimately, TRL is a Hugging Face library built on top of Transformers and is part of the Hugging Face ecosystem. If the Trainer does limit flexibility, then Transformers will need to adapt; otherwise, we will have to maintain a much larger and more complex codebase.

We'll come up with a way to add these features and prepare a PR soon!

**Comment by August-murr** [comment]
@qgallouedec, do you want to comment?

**Comment by qgallouedec** [comment]
Maybe having a `reward_func` arg of type `Callable` is an option.

Alternatively, releasing the type of `reward_model` to accept any `Callable` is also an option. But given that a custom reward func won't return the same type/shape as proper `reward_model` I'm a bit afraid that it would require overcomplicated logic.

In any case, I believe that the best approach is to discuss around a PR if anyone is willing to propose their approach

**Comment by yananchen1989** [comment]
i hear u. thanks

**Comment by yananchen1989** [comment]
i see in recent DeepSeek-R1 technical report that for GRPO, they adopt a rule-based reward system and  do not apply the outcome or process neural reward model in developing DeepSeek-R1-Zero,
because they find that the neural reward model may suffer from reward hacking in the large-scale reinforcement learning process, and retraining the reward model needs additional training resources and it complicates the whole training pipeline.

therefore, i suppose in TRL, not only for PPO, but also for GRPO and other optimization algos, thorough and complete uncoupling between the training and the reward model, may be the best practice and easy to use.

such as 
```
trainer = GRPOTrainer(
    model="Qwen/Qwen2-0.5B-Instruct",
    reward_funcs="weqweasdas/RM-Gemma-2B",
    args=training_args,
    train_dataset=dataset,
    peft_config=LoraConfig(task_type="CAUSAL_LM"),
)
```

**Comment by yananchen1989** [comment]
oh, i just see in the 0.14 release, GRPO uses this new feature, but RLOO and PPO have not been applied. 
will the `reward_funcs` be applied to all of them in future ?
@qgallouedec

**Comment by qgallouedec** [comment]
Not a priority at this point but we are open to contributions

**Comment by Superskyyy** [comment]
> oh, i just see in the 0.14 release, GRPO uses this new feature, but RLOO and PPO have not been applied. will the `reward_funcs` be applied to all of them in future ? [<img alt="" width="16" height="16" src="https://avatars.githubusercontent.com/u/45557362?u=37cea872e44a4d597f3b9c899add7509044d9ef0&amp;v=4&amp;size=80">@qgallouedec](https://github.com/qgallouedec)

RLOO PR is already merged, PPO has a draft PR, both works well. Ive tested.

**Comment by marianavargasv** [comment]
The notebook https://github.com/huggingface/trl/blob/main/examples/notebooks/gpt2-sentiment.ipynb seems to be out of date w.r.t. the current trl version `0.18.2`. I tend to agree with @yanan1116 , too much encapsulation could cause an unwanted lack of flexibility. I'm currently trying to figure out how to implement a custom reward function myself, any updates on this? Thanks!

**Comment by qgallouedec** [comment]
@shirinyamani is working on making rloo and ppo much closer to GRPO. Would it solve the issue?

**Comment by marianavargasv** [comment]
Thanks for pointing that out. Indeed I'm able to do exactly what I need with the GRPO interface. The same functionality for other algorithms would be perfect!

**Comment by qgallouedec** [comment]
Hey guys, sorry for the very slow follow-up here, catching up on the backlog.

Good news: since this thread, the GRPO-style `reward_funcs` interface (which accepts any custom callable, including a pure rule-based function) has been generalized to the other online trainers. `RLOOTrainer` now takes `reward_funcs`; same for  `OnlineDPOTrainer` (now under `trl.experimental.online_dpo`).

So the recommended practice for a rules-based reward is to pass it as a `reward_funcs` callable to RLOO or GRPO; no need to wrap it in an `nn.Module` or patch `get_reward`.

PPO itself has moved to `trl.experimental.ppo` and still expects an `nn.Module` reward_model, so for rule-based rewards I'd steer you to RLOO/GRPO rather than PPO.
